### ISSUE TITLE
G27 — Multi-Domain Adapter: Top-2 Combination

### ISSUE THREAD

**Issue by samscarrow** [git-scarrow/fast-associative-memory#39]
## Gauntlet 27 — Multi-Domain Adapter: Top-2 Combination

**Phase:** Phase 4 – Exploration
**Kill Condition:** Combined adapter trained on top-2 single-domain winners underperforms the best individual single-domain adapter on average across all evaluation datasets (in-domain + cross-domain).

**Dependencies:** G22–G25 (single-domain adapter results determine the "top-2 winners"), G26 (cross-domain baseline numbers).

---

### Objective

Create a script `benchmarks/g27_multi_adapter_top2.py` that:

1. **Determines the top-2 single-domain adapters** by reading results from G22–G25. The "winner" metric is best in-domain top-1 accuracy at the recommended vigilance operating point (v=0.80). The script should accept the top-2 domains as CLI arguments (since it can't read Notion directly).

2. **Trains a multi-domain MetricAdapter** on the combined training data of the top-2 domains:
   - `MetricAdapter(input_dim=1024, output_dim=256, hidden_dim=512)`
   - Same OHNM triplet loss protocol as `train_adapter_triplet.py`
   - 5 epochs, batch_size=128, lr=1e-4, margin=0.2
   - Training data: concatenate the DINOv2 feature caches from both winning domains
   - Save weights to `adapter_multi_top2.pt`

3. **Evaluates the multi-domain adapter on ALL evaluation datasets:**
   - **In-domain** (both top-2 domains): top-1 accuracy, prototype count, condensation ratio at v=0.80
   - **Cross-domain** (CIFAR-100 + ImageNet-R): accuracy and drop vs no-adapter baseline (same protocol as G19/G26)

4. **Computes the kill condition:**
   - Average accuracy across all evaluation datasets for the multi-domain adapter
   - Average accuracy across all evaluation datasets for the best single-domain adapter (whichever of the top-2 performed better on average)
   - If multi-domain average < best single-domain average → kill condition met

### CLI Interface

```
python benchmarks/g27_multi_adapter_top2.py \
  --domains birds cars \
  --cache-birds ./data/cub200 \
  --cache-cars ./data/stanford_cars \
  --cache-aircraft ./data/fgvc_aircraft \
  --cache-flowers ./data/flowers102 \
  --cifar-cache ./feature_cache_vitb14 \
  --imagenet-r-cache ./feature_cache_imagenet_r_vitl14 \
  --adapter-birds adapter_cub200_birds.pt \
  --adapter-cars adapter_stanford_cars.pt \
  [--epochs 5] [--batch-size 128] [--lr 1e-4] [--margin 0.2] [--seed 42] [--device auto]
```

- `--domains`: exactly 2 domain names from `{birds, cars, aircraft, flowers}` — the top-2 winners from G22–G25
- `--cache-{domain}`: path to the feature cache directory for each domain (must contain `*train*.pt` and `*test*.pt`)
- `--adapter-{domain}`: paths to the single-domain adapter weights for comparison
- Feature caches for domains not in `--domains` are still needed if they are evaluation targets

### Expected Output

```
G27 — Multi-Domain Adapter: Top-2 Combination
=========================================================================
  Dataset            Best Single   Multi-Top2   Δ (pp)
  -------------------------------------------------------------------------
  Domain A (in)      XX.XX%        XX.XX%       +X.XX
  Domain B (in)      XX.XX%        XX.XX%       +X.XX
  CIFAR-100 (cross)  XX.XX%        XX.XX%       +X.XX
  ImageNet-R (cross) XX.XX%        XX.XX%       +X.XX
  -------------------------------------------------------------------------
  AVERAGE            XX.XX%        XX.XX%       +X.XX
=========================================================================

Kill condition: Multi-top2 avg XX.XX% vs best single avg XX.XX%
→ PASS / KILL
```

### Implementation Details

- Import from existing modules: `adapter.MetricAdapter`, `fast_associative_memory.FastAssociativeMemory`, `nstp.NSTPController`
- Reuse `load_features()` from `benchmarks/gauntlet_19_cross_domain.py` for loading feature caches
- Reuse training loop from `train_adapter_triplet.py` (OHNM miner + triplet loss)
- When concatenating training data from two domains, remap labels so they don't collide (e.g., domain A labels 0..N-1, domain B labels N..N+M-1)
- Exit code 0 if pass, 1 if kill

### Files to Create
- `benchmarks/g27_multi_adapter_top2.py`

### Files to Reference (do not modify)
- `adapter.py` — `MetricAdapter` class
- `train_adapter_triplet.py` — OHNM training protocol reference
- `fast_associative_memory.py` — `FastAssociativeMemory` class
- `nstp.py` — `NSTPController` class
- `benchmarks/gauntlet_19_cross_domain.py` — feature loading + evaluation protocol
